applications/StructuralMechanicsApplication/custom_elements/prestress_membrane_element.cpp

```diff
@@ -1132,142 +1135,552 @@ void PrestressMembraneElement::CalculateMembraneElasticityTensor(
+    
+    // -->Compute contravariant basis (in total reference configuration)
+    array_1d<double, 3> base_ref_contra_tot_1, base_ref_contra_tot_2;
+    base_ref_contra_tot_1 = inv_metric_tot(0)*mG1Initial[PointNumber] + inv_metric_tot(2)*mG2Initial[PointNumber];
+    base_ref_contra_tot_2 = inv_metric_tot(2)*mG1Initial[PointNumber] + inv_metric_tot(1)*mG2Initial[PointNumber];
     
-    Tm(1, 0) = eG21*eG21;
-    Tm(1, 1) = eG22*eG22;
-    Tm(1, 2) = 2.0*eG21*eG22;
+    //-------------------------------------------
+    //--2--Compute the total deformation gradient
+    //-------------------------------------------
 
-    Tm(2, 0) = eG11*eG21;
-    Tm(2, 1) = eG12*eG22;
-    Tm(2, 2) = eG11*eG22 + eG12*eG21;
+    
+    // computation total deformation gradient
+    bounded_matrix<double,3,3> deformation_gradient_total;
+    for(unsigned int i=0; i<3; i++){
+        for(unsigned int j=0; j<3; j++){
+            deformation_gradient_total(i,j) = base_ref_contra_tot_1(j)*g1(i) + base_ref_contra_tot_2(j)*g2(i) + mG3Initial[PointNumber](j)*g3(i);
+        }
+    }
 
-    prestress = prod(Tm, prestress);
+    //-------------------------------------------
+    //--3--Compute the eigenvalues of the total deformation gradient
+    //-------------------------------------------
+    //The Right Cauchy-Green-Tensor in the local cartesian frame
+    //-> With this kind of description, only two eigenvalues are calculated
+    //lambda(3) is known, due to the fact that the thickness remains constant lambda(3)=1
 
-    //if (this->GetId() == 42 && PointNumber == 1)
-    //{
-    //    KRATOS_WATCH(Tm);
-    //    KRATOS_WATCH(prestress);
-    //}
+    // Transform the right Cauchy-Green Tensor from the total reference configuration (=gab) to the total reference cartesian basis
+    // in order to compute eigenvalues (cartesian basis necessary)
+    
+    
+    bounded_matrix<double,3,3> Origin = ZeroMatrix(3,3);
+    bounded_matrix<double,3,3> Target = ZeroMatrix(3,3);
+    bounded_matrix<double,3,3> Tensor = ZeroMatrix(3,3);
+
+
+    for (int i=0;i<3;i++){
+        Origin(i,0) = base_ref_contra_tot_1(i);
+        Origin(i,1) = base_ref_contra_tot_2(i);
+        Origin(i,2) = mG3Initial[PointNumber](i);
+        Target(i,0) = E1_tot(i);
+        Target(i,1) = E2_tot(i);
+        Target(i,2) = E3_tot(i);        
+    }
+    Tensor.clear();
+    Tensor(0,0) = gab(0);
+    Tensor(1,1) = gab(1);
+    Tensor(1,0) = gab(2);
+    Tensor(0,1) = gab(2);   
+    
+    // Transformation C Tensor
+    StructuralMechanicsMathUtilities::TensorTransformation<3>(Origin,Origin,Target,Target,Tensor);
+
+    // Compute the Eigenvalues
+    // (C-lambda_i*I)N=0
+    // Tensor:Eigenvalue in out-of-plane direction lambda_3 = 1 (no strain in this direction)
+    double lambda_1, lambda_2;
+    double root;
+    //solution quadratic formula
+    root = std::sqrt(std::abs((Tensor(0,0)+Tensor(1,1))*(Tensor(0,0)+Tensor(1,1))-4.0*(Tensor(0,0)*Tensor(1,1)-Tensor(0,1)*Tensor(1,0))));
```

Comment:
Math: No std::abs needed! (a^2 + b^2 >= 2ab)

Reply:
It better should be.. but that's no quadratic formula, isn't it? It is smaller than zero in the case that the quadratic function has no zero crossing. This should in general not be the case.. however I just added it to have the same syntax as in Carat.

Reply:
Let:
```
Tensor(0,0) = a
Tensor(1,1) = b
Tensor(0,1) = Tensor(1,0) = c
```

Then:
`root = sqrt(abs((a + b) (a + b) - 4 (a b - c ^ 2)))`

Expand:
`root = sqrt(abs(a ^ 2 + b ^ 2 + 4 c ^ 2 - 2 a c))`

Simplify with the knowledge that `a ^ 2 + b ^ 2 >= 2 a b`:

`root = sqrt(a ^ 2 + b ^ 2 + 4 c ^ 2 - 2 a c)`

Just for readability…

Reply:
done